dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -18,6 +18,150 @@
+        throw std::invalid_argument(
+                "axis " +
+                std::to_string(axis) +
+                std::string(" is out of bounds for array of dimension ") +
+                std::to_string(array_ptrs.front()->ndim()));
+    }
+
+
+    // The shape of the input arrays, which will be the
+    // same except for possibly on the concatenation axis
+    std::span<const ssize_t> shape0 = array_ptrs.front()->shape();
+    std::vector<ssize_t> shape(shape0.begin(), shape0.end());
+
+    // On the concatenation axis we sum the axis dimension sizes
+    for (auto it = std::next(array_ptrs.begin()), stop = array_ptrs.end(); it != stop; ++it) {
+       shape[axis] = shape[axis] + (*it)->shape()[axis];
+    }
+
+    return shape;
+}
+
+ConcatenateNode::ConcatenateNode(std::span<ArrayNode*> array_ptrs, const ssize_t axis)
+        : ArrayOutputMixin(make_concatenate_shape(array_ptrs, axis)), axis_(axis), array_ptrs_(array_ptrs.begin(), array_ptrs.end()) {
+
+    // Compute buffer start position for each input array
+    array_starts_.reserve(array_ptrs.size());
+    array_starts_.emplace_back(0);
+    for (ssize_t arr_i = 1, stop = array_ptrs.size(); arr_i < stop; ++arr_i) {
+        auto subshape = array_ptrs_[arr_i - 1]->shape().last(this->ndim() - axis_);
+        ssize_t prod = std::accumulate(subshape.begin(), subshape.end(), 1, std::multiplies<ssize_t>());
+        array_starts_.emplace_back(prod + array_starts_[arr_i - 1]);
+    }
+
+    for (auto it = array_ptrs.begin(), stop = array_ptrs.end(); it != stop; ++it) {
+        this->add_predecessor((*it));
+    }
+}
+
+double const* ConcatenateNode::buff(const State& state) const {
+    return data_ptr<ArrayNodeStateData>(state)->buff();
+}
+
+std::span<const Update> ConcatenateNode::diff(const State& state) const {
+   return data_ptr<ArrayNodeStateData>(state)->diff();
+}
+
+void ConcatenateNode::initialize_state(State& state) const {
+    int index = topological_index();
+    assert(index >= 0 && "must be topologically sorted");
+    assert(static_cast<int>(state.size()) > index && "unexpected state length");
+    assert(state[index] == nullptr && "already initialized state");
+
+    std::vector<double> values;
+    values.resize(size());
+
+    for (ssize_t arr_i = 0, stop = array_ptrs_.size(); arr_i < stop; ++arr_i) {
+        // Create a view into our buffer with the same shape as
+        // our input array starting at the correct place
+        auto view_it = ArrayIterator(
+                            values.data() + array_starts_[arr_i],
+                            this->ndim(),
+                            array_ptrs_[arr_i]->shape().data(),
+                            this->strides().data());
+
+        std::copy(array_ptrs_[arr_i]->begin(state), array_ptrs_[arr_i]->end(state), view_it);
+    }
+
+    state[index] = std::make_unique<ArrayNodeStateData>(std::move(values));
+}
+
+void ConcatenateNode::commit(State& state) const {
```

Comment:
This is an aesthetic thing, and won't block the PR, but I prefer the methods to be in alphabetical order. I know some of the files don't have that - it's legacy code that will be updated at some later point.

Reply:
Fixed so most of the methods are now ordered alphabetically. 
Is there a way to forward declare the `make_concatenate_shape`? If not it's forced to be defined before the constructors

Reply:
Fixed with [7693a54](https://github.com/dwavesystems/dwave-optimization/pull/153/commits/7693a54039726aabb6fa0cb1ec0c3a656de8f53b)